cassiopeia/data/utilities.py

```diff
@@ -119,6 +119,77 @@ def _to_newick_str(g, node):
     return _to_newick_str(tree, root) + ";"
 
 
+def remove_and_prune_lineage(node: int, tree: nx.DiGraph) -> None:
+    """Removes a node and prunes the lineage.
+
+    Removes a node and all ancestors of that node that are no longer the
+    ancestor of any leaves. In the context of a lineage tracing
+    experiment, this removes all ancestral nodes that are not the
+    ancestors of any observed samples at the end of the experiment, thus
+    pruning all lineages that died.
+
+    Args:
+        node: The node to be removed
+        tree: The tree to remove the node from
+    """
+    if len(tree.nodes) > 1:
+        curr_parent = list(tree.predecessors(node))[0]
+        tree.remove_node(node)
+        while (
+            tree.out_degree(curr_parent) < 1 and tree.in_degree(curr_parent) > 0
+        ):
+            next_parent = list(tree.predecessors(curr_parent))[0]
+            tree.remove_node(curr_parent)
+            curr_parent = next_parent
+
+
+def collapse_unifurcations(tree: nx.DiGraph, source: int = None) -> None:
+    """Collapses unifurcations in a given tree.
+
+    Args:
+        tree: The tree to collapse unifurcations on
+        source: The node at which to begin the tree traversal
+    """
+
+    weighted = True
+    for u, v in tree.edges:
+        if not tree.get_edge_data(u, v):
+            weighted = False
+
+    def _collapse_unifurcations(tree, node, parent):
+        succs = list(tree.successors(node))
+        if len(succs) == 1:
+            if weighted:
+                t = tree.get_edge_data(parent, node)["weight"]
+                t_ = tree.get_edge_data(node, succs[0])["weight"]
+                tree.add_edge(parent, succs[0], weight=t + t_)
+            else:
+                tree.add_edge(parent, succs[0])
+            _collapse_unifurcations(tree, succs[0], parent)
+            tree.remove_node(node)
+        else:
+            for i in succs:
+                _collapse_unifurcations(tree, i, node)
+
+    if not source:
+        source = [n for n in tree if tree.in_degree(n) == 0][0]
+
+    for node in tree.successors(source):
+        _collapse_unifurcations(tree, node, source)
+
+    succs = list(tree.successors(source))
+    if len(succs) == 1:
+        if weighted:
+            t = tree.get_edge_data(source, succs[0])["weight"]
+            for i in tree.successors(succs[0]):
+                t_ = tree.get_edge_data(succs[0], i)["weight"]
+                tree.add_edge(source, i, weight=t + t_)
+        else:
+            for i in tree.successors(succs[0]):
+                tree.add_edge(source, i)
+        tree.remove_node(succs[0])
+
+
```

Comment:
One remark is that if we perform this operation in the CassiopeiaTree class, we will have to be careful about updating the branch lengths. I think it could be useful to have the collapse unifurcations functionality in a CassiopeiaTree too, but I think a version for networkx is also useful to have. To not duplicate code, I would imagine the CassiopeiaTree implementation leveraging the networkX implementation.

I lean towards accepting the current code and leaving the CassiopeiaTree version as a TODO, just because I think other things might be of higher priority.